### PR TITLE
Mask sensitive query parameters in request logs

### LINKED ISSUE #6919: Provide a way to mask sensitive query parameters in request logs
The default `LogFormatter` logs the full `:path` pseudo header, which includes the query string. If an application receives sensitive values as query parameters, for example `/v1/users?token=abcdef`, they are written to the log as is. `HeadersSanitizer` can mask a header value, but there is no way to mask a single query parameter.

The workaround today is to treat `:path` as a sensitive header and rewrite it by hand:

```java
HeadersSanitizer.builderForText()
                .sensitiveHeaders(HttpHeaderNames.AUTHORIZATION, HttpHeaderNames.COOKIE,
                                  HttpHeaderNames.PATH)
                .maskingFunction((name, value) -> {
                    if (!name.equals(HttpHeaderNames.PATH)) {
                        return "****";
                    }
                    // Parse, mask and re-encode the query string here.
                    ...
                })
                .build();
```

This has a few problems:
- The user has to parse and re-encode the query string, even though `QueryParams` already does this.
- Masking is all or nothing per header, so it is hard to mask only `token` while keeping `page=1`.
- `sensitiveHeaders()` replaces the default set, so the caller must remember to re-add `Authorization`, `Cookie` and the others.

It would be useful if the query parameters to mask could be specified directly:

```java
LogFormatter.builderForText()
            // Mask the specified query parameters in the request path.
            .maskQueryParams("token", "ssn")
            // Or provide a custom function.
            .queryParamMaskingFunction((name, value) -> "****")
            .build();
```

### PR DESCRIPTION
**Pull request by poketopa**
Motivation:

`LogFormatter` currently logs sensitive query parameter values as part of the `:path` pseudo-header. Masking only selected query parameters requires users to replace the request header sanitizer and rewrite the request path themselves, which can also replace Armeria's default sensitive-header masking behavior.

Modifications:

- Add `maskQueryParams(String...)` and `maskQueryParams(Iterable<String>)` to the text and JSON log formatter builders.
- Add `QueryParamMaskingFunction` and `queryParamMaskingFunction(...)` for custom replacement or removal of selected query parameters.
- Apply query parameter masking before the existing request header sanitizer so default and custom sanitizers continue to work.
- Preserve unmasked raw query components, encoded parameter names, fragments, duplicate parameters and their order.
- Add regression tests for text and JSON formatting, duplicate parameters, custom masking and removal, raw component preservation, and sanitizer composition.

Result:

- Closes #6919.
- Applications can mask selected query parameter values in request logs without replacing Armeria's existing sensitive-header masking behavior.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/logging/AbstractLogFormatterBuilder.java [hunk 1]
```diff
@@ -53,6 +62,49 @@ final SELF self() {
...
+
+    /**
+     * Sets the {@link QueryParamMaskingFunction} to use to mask query parameters before logging.
+     * The default is {@link QueryParamMaskingFunction#of()}.
+     */
+    public SELF queryParamMaskingFunction(QueryParamMaskingFunction queryParamMaskingFunction) {
+        this.queryParamMaskingFunction =
+                requireNonNull(queryParamMaskingFunction, "queryParamMaskingFunction");
+        return self();
+    }
```
[the remark this answers is not in the document]
Comment (poketopa) [reply core/src/main/java/com/linecorp/armeria/common/logging/AbstractLogFormatterBuilder.java:100]: Added @UnstableApi to QueryParamMaskingFunction and all three builder methods in 486c258.

File: core/src/main/java/com/linecorp/armeria/common/logging/QueryParamMaskingHeadersSanitizer.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    @Override
+    public T sanitize(RequestContext ctx, HttpHeaders headers) {
+        final String path = headers.get(HttpHeaderNames.PATH);
+        if (path == null) {
+            return delegate.sanitize(ctx, headers);
+        }
+
+        final int queryStart = path.indexOf('?') + 1;
+        if (queryStart == 0 || queryStart == path.length()) {
+            return delegate.sanitize(ctx, headers);
```
[the remark this answers is not in the document]
Comment (poketopa) [reply core/src/main/java/com/linecorp/armeria/common/logging/QueryParamMaskingHeadersSanitizer.java]: Updated query delimiter handling so a question mark after a fragment is not treated as a query, and added a raw RequestHeaders regression test in 486c258.

File: core/src/main/java/com/linecorp/armeria/common/logging/QueryParamMaskingHeadersSanitizer.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        if (maskedPath == null) {
+            return delegate.sanitize(ctx, headers);
+        }
+        return delegate.sanitize(
+                ctx, headers.withMutations(builder -> builder.set(HttpHeaderNames.PATH, maskedPath)));
+    }
+
+    @Nullable
+    private String maskQueryParams(String path, int queryStart) {
+        final int fragmentStart = path.indexOf('#', queryStart);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/logging/QueryParamMaskingHeadersSanitizer.java:70]: nit:
```suggestion
    private String maskQueryParams(String path, int queryStart, int fragmentStart) {
```

File: core/src/main/java/com/linecorp/armeria/common/logging/QueryParamMaskingHeadersSanitizer.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        final int queryEnd = fragmentStart >= 0 ? fragmentStart : path.length();
+        if (queryStart == queryEnd) {
+            return null;
+        }
+
+        try (TemporaryThreadLocals tempThreadLocals = TemporaryThreadLocals.acquire()) {
+            final StringBuilder maskedQuery = tempThreadLocals.stringBuilder();
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/logging/QueryParamMaskingHeadersSanitizer.java:77]: I think we can create just one `StringBuilder`:
```
final StringBuilder maskedPath = tempThreadLocals.stringBuilder();
maskedPath.append(path, 0, queryStart);
....

if (!retained) {
  maskedPath.setLength(queryStart - 1);
}
```

**Review by minwoox — COMMENTED** [review]
Looks good to me 👍

File: core/src/main/java/com/linecorp/armeria/common/logging/QueryParamMaskingHeadersSanitizer.java (hunk 4 above)
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/logging/QueryParamMaskingHeadersSanitizer.java:77]: Have you considered using `QueryParams` instead of parsing it yourself?
`QueryParams.fromQueryString(query, Integer.MAX_VALUE)`

**Comment by jrhee17** [comment]
Thanks for the PR.

I'm wondering if it makes more sense if we allow users to specify multiple `HeaderMaskingFunction`s on a `AbstractHeadersSanitizerBuilder`. By doing so, we could 1) let `queryParamMaskingFunction` be a `HeaderMaskingFunction`, and this can give users freedom to specify more masking functions for structured headers.

```
private Map<AsciiString, List<HeaderMaskingFunction>> perHeaderMaskingFunctions;

public SELF maskHeader(CharSequence headerName, HeaderMaskingFunction maskingFunction) {
    ...
}

// later
final Map<AsciiString, List<HeaderMaskingFunction>> headerMaskingFunctions() {
    // 1. compute based on maskingFunction, sensitiveHeaders
    // 2. compute based on queryParamMaskingFunction
    // 3. user-specified per-header custom functions.
}
```

Also, we can add convenience methods for query params as well:

```
public SELF maskQueryParams(Iterable<String> queryParams) {
    // set queryParamMaskingFunction as queryParams -> "****"
}

public SELF queryParamMaskingFunction(QueryParamMaskingFunction queryParamMaskingFunction) {
    this.queryParamMaskingFunction =
            requireNonNull(queryParamMaskingFunction, "queryParamMaskingFunction");
    return self();
}
```

cc. @line/dx since this comment can change the direction of the PR

**Comment by jrhee17** [comment]
ping @poketopa 🙇

**Comment by ikhoon** [comment]
> if we allow users to specify multiple HeaderMaskingFunctions on a AbstractHeadersSanitizerBuilder

That sounds like a good idea. Alternatively, we could provide a builder for `HeaderMaskingFunction` so that sanitizers for different types of headers can be configured fluently. It would be even better if multiple functions could be composed using something like `orElse()`.

```java
HeaderMaskingFunction
    .builder()
    .onAuthorization(token -> "****")
    .onQueryParam((paramName, value) -> ...)
    .onCookie(cookie -> ...)
    .onHeader((name, value) -> ...)
    .build();
```

**Comment by jrhee17** [comment]
> Alternatively, we could provide a builder for HeaderMaskingFunction so that sanitizers for different types of headers can be configured fluently. 

Sounds fine as long as the relationship between `sensitiveHeaders` and `HeaderMaskingFunction` is defined

**Comment by jrhee17** [comment]
ping @poketopa 🙇
